### PR TITLE
Vyper scaffolding

### PR DESCRIPTION
**Pull request by internnos**
## Motivation

Added support for Vyper

### CODE REVIEW

File: brownie-config.yaml
Comment (0age) [inline brownie-config.yaml:2]: Could we put everything vyper-related into a vyper root folder?

File: brownie-config.yaml
Comment (0age) [inline brownie-config.yaml:38]: Does this imply to use the latest version?

File: build/contracts/Greeter.json
Comment (0age) [inline build/contracts/Greeter.json:1]: Let’s put the vyper build folder in .gitignore

File: build/tests.json
Comment (0age) [inline build/tests.json:1]: Same, gitignore’d please

**Review by 0age — CHANGES_REQUESTED** [review]
Looking good so far, left some comments

File: brownie-config.yaml
Comment (internnos) [reply brownie-config.yaml:2]: ofc, including the tests and contracts, or just the brownie-related files?

File: brownie-config.yaml
Comment (internnos) [reply brownie-config.yaml:38]: nope ser, it actually depends on the pragma used in the contract

This is taken from the [documentation](https://eth-brownie.readthedocs.io/en/stable/config.html#config-reports)

> The version of vyper to use. Should be given as a string in the format 0.x.x. If set to null, the version is set based on the contract pragma

File: build/contracts/Greeter.json
Comment (internnos) [reply build/contracts/Greeter.json:1]: agree, on it

File: build/tests.json
Comment (internnos) [reply build/tests.json:1]: on it

File: brownie-config.yaml
Comment (0age) [reply brownie-config.yaml:2]: Let’s do it all

File: brownie-config.yaml
Comment (0age) [reply brownie-config.yaml:38]: Perfect

File: brownie-config.yaml
Comment (internnos) [reply brownie-config.yaml:2]: like this?

```
project_structure:
    build: vyper/build
    contracts: vyper/contracts/
    interfaces: vyper/interfaces
    reports: vyper/reports
    scripts: vyper/scripts
    tests: vyper/test
```

Comment (0age) [reply brownie-config.yaml:2]: Exactly

File: brownie-config.yaml
Comment (0age) [inline brownie-config.yaml:1]: Oh one other thing: can we put this file in the `config/` folder and specify it’s location via a flag somehow? Not a big deal if we can’t but want to keep the root level uncluttered where possible

File: brownie-config.yaml
Comment (internnos) [reply brownie-config.yaml]: done

File: brownie-config.yaml
Comment (internnos) [reply brownie-config.yaml:1]: oh yeah i was thinking about that too, but i haven't been able to do it until now

this is from the [documentation](https://eth-brownie.readthedocs.io/en/stable/config.html#config-reports)

> The configuration file must be saved as brownie-config.yaml. If saved in the root directory of a project it will be loaded whenever that project is active. If saved in your [home path](https://docs.python.org/3/library/pathlib.html#pathlib.Path.home), it will always be loaded.


i'll keep trying, though

**Review by 0age — APPROVED** [review]
LGTM!

**Comment by banteg** [comment]
no need to copy over the entire brownie config, a usual config should be pretty minimal, something like:
```
networks:
    default: hardhat

project_structure:
    build: vyper/build
    contracts: vyper/contracts/
    interfaces: vyper/interfaces
    reports: vyper/reports
    scripts: vyper/scripts
    tests: vyper/test

autofetch_sources: true
```

if you already use hardhat, there is no need introduce another dependency of ganache, brownie supports hardhat properly.

for pytest+foundry, we'd likely need to migrate to ape later, but hold off for now, it doesn't support everything we need yet.

**Comment by internnos** [comment]
> ### config
> no need to copy over the entire brownie config, a usual config should be pretty minimal, something like:
> 
> ```yaml
> networks:
>     default: hardhat
> 
> project_structure:
>     build: vyper/build
>     contracts: vyper/contracts/
>     interfaces: vyper/interfaces
>     reports: vyper/reports
>     scripts: vyper/scripts
>     tests: vyper/test
> 
> autofetch_sources: true
> ```
> 
> ### hardhat
> if you already use hardhat, there is no need introduce another dependency of ganache, brownie supports hardhat properly.
> 
> for pytest+foundry, we'd likely need to migrate to ape later, but hold off for now, it doesn't support everything we need yet.
> 
> ### isolation
> also add a vyper/test/conftest.py with this to get the test chain reverted to a state after all fixtures are set up after each test:
> 
> ```python
> import pytest
> 
> @pytest.fixture(autouse=True)
> def isolate(fn_isolation):
>     pass
> ```
> 
> read more [here](https://eth-brownie.readthedocs.io/en/stable/tests-pytest-intro.html#pytest-fixtures-isolation)
> 
> ### dev revert comments
> also note that brownie supports [dev revert strings](https://eth-brownie.readthedocs.io/en/stable/tests-pytest-intro.html#developer-revert-comments), e.g.
> 
> ```python
> assert water > 0, "no water in seaport"
> ```
> 
> can be written as a comment
> 
> ```python
> assert water > 0  # dev: no water in seaport
> ```
> 
> then it can be caught in tests like this:
> 
> ```python
> with brownie.reverts(dev_revert_msg="no water in seaport"):
>     seaport.check_water_levels()
> ```

gotcha, on it ser

**Review by banteg — APPROVED** [review]

**Review by storming0x — APPROVED** [review]
Changes LGTM, as part of this scope do you want to setup CI github actions?

this is one example we have, altho it may have too many steps not needed at this point.

https://github.com/yearn/yearn-vaults/blob/main/.github/workflows/test.yaml

**Comment by 0age** [comment]
> Changes LGTM, as part of this scope do you want to setup CI github actions?
> 
> this is one example we have, altho it may have some steps that arent needed at this point.
> 
> https://github.com/yearn/yearn-vaults/blob/main/.github/workflows/test.yaml

Let’s get this in on the next PR so that we unblock on setting up scaffolding!

**Pull request merged**
